runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -4758,10 +4763,33 @@ J9::Z::TreeEvaluator::multianewArrayEvaluator(TR::Node * node, TR::CodeGenerator
 
    bool use64BitClasses = comp->target().is64Bit() && !TR::Compiler->om.generateCompressedObjectHeaders();
 
-   // Init class field, then jump to end of ICF
    generateRXInstruction(cg, use64BitClasses ? TR::InstOpCode::STG : TR::InstOpCode::ST, node, classReg, generateS390MemoryReference(targetReg, TR::Compiler->om.offsetOfObjectVftField(), cg));
-   cursor = generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BRC, node, cFlowRegionEnd);
-   iComment("Init class field and jump to end of ICF.");
+#if defined(TR_TARGET_64BIT)
+   if (isIndexableDataAddrPresent)
+      {
+      TR_ASSERT_FATAL_WITH_NODE(node,
+         (TR::Compiler->om.compressObjectReferences()
+               && (fej9->getOffsetOfDiscontiguousDataAddrField() - fej9->getOffsetOfContiguousDataAddrField()) == 8)
+            || (!TR::Compiler->om.compressObjectReferences()
+               && fej9->getOffsetOfDiscontiguousDataAddrField() == fej9->getOffsetOfContiguousDataAddrField()),
+         "Offset of dataAddr field in discontiguous array is expected to be 8 bytes more than contiguous array if using compressed refs, "
+         "Or same if using full refs. But was %d bytes for discontigous and %d bytes for contiguous array.\n",
+         fej9->getOffsetOfDiscontiguousDataAddrField(), fej9->getOffsetOfContiguousDataAddrField());
+
+      // Load dataAddr slot offset difference since 0 size arrays are treated as discontiguous.
+      generateRIInstruction(cg,
+         TR::InstOpCode::LGHI,
+         node,
+         temp1Reg,
+         static_cast<int32_t>(fej9->getOffsetOfDiscontiguousDataAddrField() - fej9->getOffsetOfContiguousDataAddrField()));
+      cursor = generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BRC, node, populateFirstDimDataAddrSlot);
+      }
+   else
+#endif /* TR_TARGET_64BIT */
+      {
+      cursor = generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BRC, node, cFlowRegionEnd);
```

Comment:
We would not have a branch to `cFlowRegionEnd` generated when indexable data addr is set. can  we should be able to refactor this to avoid two `generateS390BranchInstruction` especially when both are unconditional branch ? Also, I would carefully check the Internal Control Flow region as well.

Reply:
> We would not have a branch to cFlowRegionEnd generated when indexable data addr is set.

That is intentional. `populateFirstDimDataAddrSlot` is right before `cFlowRegionEnd` so when dataAddr field is toggled on, we jump to `populateFirstDimDataAddrSlot` and then go to `cFlowRegionEnd` instead of jumping straight to `cFlowRegionEnd`.

Reply:
Do we ever have instruction that branches to cFlowRegionEnd when indexable data addr is set?

Reply:
Nope, we always jump to `populateFirstDimDataAddrSlot` then go to `cFlowRegionEnd` when dataAddr field is toggled on.

Reply:
Ok in that case, atleast what I can understand from the evaluator (we should verify this through the log file as well ), you do not need to generate `populateFirstDimDataAddrSlot` label symbol, and generate `cFlowRegionEnd` before you populate dataAddrSlot at line 4872. Based on your comment in https://github.com/eclipse-openj9/openj9/pull/16420/files#r1166123941, your ICF ends at `populateFirstDimDataAddrSlot` in case of `isIndexableDataAddrPresent` is set.

Again, this is simply based on the previous comment, I would like to take a look at the log file and verify this as well.

Reply:
Ok, in that case, you may even do not need to generate `populateFirstDimDataAddrSlot`. You should be able to generate `cFlowRegionEnd` before you populate the data Addr slot as that is where your ICF ends (This is purely based on what I can understand from the evaluator and your last comment, we should verify this through a log file for sanity as well).